semantics-3.0/Semantics.hs

```diff
@@ -1,34 +1,38 @@
-{-# LANGUAGE NamedFieldPuns #-}
 module Semantics where
 
-import           Data.List       (foldl')
 import           Data.Map.Strict (Map)
 import qualified Data.Map.Strict as Map
-import           Data.Set        (Set)
-import qualified Data.Set        as Set
 
-type SlotNumber = Integer
-type SlotInterval = (SlotNumber, SlotNumber)
+newtype Slot = Slot { getSlot :: Integer } deriving (Eq,Ord)
+
+instance Show Slot where
+    show (Slot n) = "(Slot " ++ show n ++ ")"
+
+newtype Ada = Lovelace { getLovelace :: Integer } deriving (Eq,Ord)
```

Comment:
How about deriving some Number classes, instead of `getLovelace first - getLovelace second` you have `first - second`

Reply:
Done.